File: static/js/job-parser-api.js
/**
 * JobParserAPI - Classe pour l'intégration du parser de fiches de poste
 * Cette classe fournit des méthodes pour analyser des fiches de poste
 * à partir de fichiers PDF ou de texte brut.
 */
class JobParserAPI {
    /**
     * Initialise l'API du parser de fiches de poste
     * @param {Object} config - Configuration de l'API
     * @param {boolean} config.debug - Active le mode debug
     * @param {boolean} config.enablePDFCleaning - Active le nettoyage des PDF
     * @param {string} config.apiUrl - URL de l'API (par défaut: détection automatique ou localhost)
     * @param {boolean} config.forceLocalProcessing - Force l'analyse locale sans appeler l'API
     */
    constructor(config = {}) {
        // Récupérer l'URL de l'API depuis les paramètres de l'URL si disponible
        const urlParams = new URLSearchParams(window.location.search);
        const apiUrlFromParams = urlParams.get('apiUrl');
        
        this.config = {
            debug: config.debug || urlParams.has('debug'),
            enablePDFCleaning: config.enablePDFCleaning || false,
            apiUrl: config.apiUrl || apiUrlFromParams || this._detectApiUrl(),
            forceLocalProcessing: config.forceLocalProcessing || urlParams.has('localOnly')
        };
        
        if (this.config.debug) {
            console.log('JobParserAPI initialisée avec la configuration:', this.config);
        }
    }
    
    /**
     * Détecte automatiquement l'URL de l'API en fonction de l'environnement
     * @returns {string} URL de l'API
     * @private
     */
    _detectApiUrl() {
        // Essayer de détecter l'adresse du serveur en fonction de l'URL courante
        const currentUrl = new URL(window.location.href);
        const hostname = currentUrl.hostname;
        
        if (hostname === 'localhost' || hostname === '127.0.0.1') {
            return 'http://localhost:5055';
        } else if (hostname.includes('github.io')) {
            // En mode GitHub Pages, tenter plusieurs URLs connues
            return window.location.protocol + '//' + 
                  (window.location.hostname === 'localhost' ? 'localhost' : window.location.hostname) + 
                  ':5055';
        } else {
            // Fallback sur l'adresse IP locale ou le nom d'hôte actuel
            return window.location.protocol + '//' + hostname + ':5055';
        }
    }
    
    /**
     * Analyse un fichier PDF de fiche de poste
     * @param {File} file - Fichier PDF à analyser
     * @returns {Promise<Object>} - Résultat de l'analyse
     */
    async parseJobFile(file) {
        if (this.config.debug) {
            console.log('Analyse du fichier:', file.name);
        }
        
        // Si on force le traitement local, ne pas essayer l'API
        if (this.config.forceLocalProcessing) {
            if (this.config.debug) {
                console.log('Mode d\'analyse locale forcé, pas d\'appel API');
            }
            const text = await this._extractTextFromFile(file);
            return this.analyzeJobLocally(text);
        }
        
        try {
            const formData = new FormData();
            formData.append('job_post_file', file);
            
            if (this.config.debug) {
                console.log('Tentative d\'appel API à:', this.config.apiUrl);
            }
            
            try {
                // Vérifier si l'API est accessible
                const pingResponse = await fetch(`${this.config.apiUrl}/`, {
                    method: 'GET',
                    mode: 'cors',
                    cache: 'no-cache',
                    headers: {
                        'Accept': 'application/json'
                    },
                    timeout: 2000 // Timeout court pour le ping
                }).catch(e => {
                    if (this.config.debug) {
                        console.warn('Erreur lors du ping de l\'API:', e);
                    }
                    throw new Error('API indisponible');
                });
                
                if (!pingResponse || !pingResponse.ok) {
                    throw new Error('API indisponible');
                }
            } catch (pingError) {
                if (this.config.debug) {
                    console.warn('API indisponible, fallback sur l\'analyse locale', pingError);
                }
                const text = await this._extractTextFromFile(file);
                return this.analyzeJobLocally(text);
            }
            
            const response = await fetch(`${this.config.apiUrl}/api/parse-job-post`, {
                method: 'POST',
                mode: 'cors',
                cache: 'no-cache',
                body: formData
            });
            
            if (!response.ok) {
                throw new Error(`Erreur HTTP: ${response.status}`);
            }
            
            const result = await response.json();
            
            if (this.config.debug) {
                console.log('Résultat de l\'analyse API:', result);
            }
            
            if (result.success) {
                return result.data;
            } else {
                throw new Error(result.error || 'Erreur lors de l\'analyse');
            }
        } catch (error) {
            if (this.config.debug) {
                console.error('Erreur lors de l\'analyse du fichier via API:', error);
            }
            
            // En cas d'erreur avec l'API, essayer l'analyse locale
            if (this.config.debug) {
                console.log('Fallback sur l\'analyse locale');
            }
            const text = await this._extractTextFromFile(file);
            return this.analyzeJobLocally(text);
        }
    }
    
    /**
     * Analyse un texte de fiche de poste
     * @param {string} text - Texte de la fiche de poste
     * @returns {Promise<Object>} - Résultat de l'analyse
     */
    async parseJobText(text) {
        if (this.config.debug) {
            console.log('Analyse du texte de la fiche de poste');
        }
        
        // Si on force le traitement local, ne pas essayer l'API
        if (this.config.forceLocalProcessing) {
            if (this.config.debug) {
                console.log('Mode d\'analyse locale forcé, pas d\'appel API');
            }
            return this.analyzeJobLocally(text);
        }
        
        try {
            try {
                // Vérifier si l'API est accessible
                const pingResponse = await fetch(`${this.config.apiUrl}/`, {
                    method: 'GET',
                    mode: 'cors',
                    cache: 'no-cache',
                    headers: {
                        'Accept': 'application/json'
                    },
                    timeout: 2000 // Timeout court pour le ping
                }).catch(e => {
                    if (this.config.debug) {
                        console.warn('Erreur lors du ping de l\'API:', e);
                    }
                    throw new Error('API indisponible');
                });
                
                if (!pingResponse || !pingResponse.ok) {
                    throw new Error('API indisponible');
                }
            } catch (pingError) {
                if (this.config.debug) {
                    console.warn('API indisponible, fallback sur l\'analyse locale', pingError);
                }
                return this.analyzeJobLocally(text);
            }
            
            const response = await fetch(`${this.config.apiUrl}/api/parse-job-post`, {
                method: 'POST',
                mode: 'cors',
                cache: 'no-cache',
                headers: {
                    'Content-Type': 'application/json',
                    'Accept': 'application/json'
                },
                body: JSON.stringify({ job_post_text: text })
            });
            
            if (!response.ok) {
                throw new Error(`Erreur HTTP: ${response.status}`);
            }
            
            const result = await response.json();
            
            if (this.config.debug) {
                console.log('Résultat de l\'analyse API:', result);
            }
            
            if (result.success) {
                return result.data;
            } else {
                throw new Error(result.error || 'Erreur lors de l\'analyse');
            }
        } catch (error) {
            if (this.config.debug) {
                console.error('Erreur lors de l\'analyse du texte via API:', error);
            }
            
            // En cas d'erreur avec l'API, essayer l'analyse locale
            return this.analyzeJobLocally(text);
        }
    }
    
    /**
     * Analyse locale d'une fiche de poste (fallback si l'API n'est pas disponible)
     * @param {string} text - Texte de la fiche de poste
     * @returns {Object} - Résultat de l'analyse
     */
    analyzeJobLocally(text) {
        if (this.config.debug) {
            console.log('Analyse locale du texte de la fiche de poste');
        }
        
        // Extraction du titre du poste
        const titlePatterns = [
            /\b(?:poste|position|intitulé du poste|titre du poste|offre d'emploi)\s*[:-]\s*([^\.;\n\r]+)/i,
            /\brecherch(?:e|ons|ant)\s+(?:un(?:e)?)\s+([^\.;\n\r]+)/i,
            /\b(?:nous|entreprise)\s+recrut(?:e|ons)\s+(?:un(?:e)?)\s+([^\.;\n\r]+)/i,
            /\b((?:développeur|ingénieur|technicien|consultant|manager|responsable|directeur|analyste)[\s\w-]+)/i
        ];
        
        // Extraction de l'entreprise
        const companyPatterns = [
            /\b(?:entreprise|société|employeur|pour le compte de|au sein de)\s*[:-]\s*([^\.;\n\r]+)/i,
            /\b([\w\s-]+)\s+(?:est|recherche|recrute|souhaite)/i
        ];
        
        // Extraction de la localisation
        const locationPatterns = [
            /\b(?:lieu|localisation|localité|ville|site|basé à|poste basé|poste situé)\s*[:-]\s*([^\.;\n\r]+)/i,
            /\bà\s+([\w\s-]+)\s+\(\d{2,5}\)/i,
            /\b([\w\s-]+)\s+\(\d{2,5}\)/i
        ];
        
        // Extraction du type de contrat
        const contractPatterns = [
            /\b(?:type de contrat|contrat|type d'emploi)\s*[:-]\s*([^\.;\n\r]+)/i,
            /\b(CDI|CDD|Stage|Alternance|Intérim|Freelance|temps plein|temps partiel|permanent|temporaire)/i
        ];
        
        // Extraction de l'expérience requise
        const experiencePatterns = [
            /\b(?:expérience|années d'expérience|niveau d'expérience)\s*[:-]\s*([^\.;\n\r]+)/i,
            /\b(\d+[\s-]+ans?\s+d'expérience)/i,
            /\bexpérience\s+(?:de|d'au moins|minimum|min\.|requise|souhaitée)\s+(\d+[\s-]+ans?)/i,
            /\b(débutant|junior|confirmé|sénior|senior|expert)/i
        ];
        
        // Recherche de compétences techniques courantes dans le texte
        const commonSkills = [
            'JavaScript', 'React', 'Angular', 'Vue.js', 'Node.js', 'Python', 'Django', 'Flask',
            'Java', 'Spring', 'C#', '.NET', 'PHP', 'Laravel', 'Symfony', 'Ruby', 'Rails',
            'Go', 'Rust', 'TypeScript', 'HTML', 'CSS', 'SASS', 'LESS', 'SQL', 'MySQL',
            'PostgreSQL', 'MongoDB', 'Redis', 'Elasticsearch', 'Docker', 'Kubernetes',
            'AWS', 'Azure', 'GCP', 'CI/CD', 'Jenkins', 'Git', 'Agile', 'Scrum', 'Jira',
            'DevOps', 'TDD', 'Swift', 'Kotlin', 'Unity', 'Flutter', 'Dart', 'Firebase'
        ];
        
        // Recherche de formations/diplômes courants
        const educationPatterns = [
            /\b(?:formation|diplôme|niveau d'études|niveau de formation|profil académique)\s*[:-]\s*([^\.;\n\r]+)/i,
            /\b(Bac\s*\+\s*\d|Master|Licence|Bachelor|Ingénieur|Doctorat|MBA|BTS|DUT)/i
        ];
        
        // Recherche de salaire/rémunération
        const salaryPatterns = [
            /\b(?:salaire|rémunération|package|traitement|appointements)\s*[:-]\s*([^\.;\n\r]+)/i,
            /\b(\d+[\s-]*[kK€]|\d+[\s-]*000[\s-]*€|\d+[\s-]*000[\s-]*euros)/i
        ];
        
        // Extraction des informations à partir des patterns définis
        const extractInfo = (patterns, defaultValue = 'Not specified') => {
            for (const pattern of patterns) {
                const match = text.match(pattern);
                if (match && match[1] && match[1].trim()) {
                    return match[1].trim();
                }
            }
            return defaultValue;
        };
        
        // Extraction des compétences techniques
        const extractSkills = () => {
            const skills = [];
            for (const skill of commonSkills) {
                const pattern = new RegExp('\\b' + skill + '\\b', 'i');
                if (pattern.test(text)) {
                    skills.push(skill);
                }
            }
            return skills.length > 0 ? skills : ['Not specified'];
        };
        
        // Analyse des responsabilités/missions
        const extractResponsibilities = () => {
            const responsibilitiesPatterns = [
                /(?:missions|responsabilités|tâches|activités|missions principales)[\s\n]*:[\s\n]*((?:[\s\n]*[-•*][\s\n]*[^;:\.]*[;:\.]*[\s\n]*)+)/i,
                /(?:vous serez chargé de|vous serez responsable de|vous serez amené à)[\s\n]*:[\s\n]*((?:[\s\n]*[-•*][\s\n]*[^;:\.]*[;:\.]*[\s\n]*)+)/i
            ];
            
            for (const pattern of responsibilitiesPatterns) {
                const match = text.match(pattern);
                if (match && match[1]) {
                    // Extraction des items à puces
                    const itemsText = match[1];
                    const items = itemsText.split(/[\s\n]*[-•*][\s\n]*/).filter(item => item.trim().length > 0);
                    if (items.length > 0) {
                        return items.map(item => item.trim());
                    }
                }
            }
            return ['Not specified'];
        };
        
        // Extraction des avantages
        const extractBenefits = () => {
            const benefitsPatterns = [
                /(?:avantages|benefits|nous offrons|nous proposons|nous vous offrons|package|ce que nous offrons)[\s\n]*:[\s\n]*((?:[\s\n]*[-•*][\s\n]*[^;:\.]*[;:\.]*[\s\n]*)+)/i
            ];
            
            const commonBenefits = [
                'tickets restaurant', 'mutuelle', 'prévoyance', 'RTT', 'télétravail', 'remote',
                'flexible', 'formation', 'transport', 'treizième mois', '13ème mois',
                'bonus', 'intéressement', 'participation', 'CE', 'comité d\'entreprise'
            ];
            
            // Chercher par pattern
            for (const pattern of benefitsPatterns) {
                const match = text.match(pattern);
                if (match && match[1]) {
                    const itemsText = match[1];
                    const items = itemsText.split(/[\s\n]*[-•*][\s\n]*/).filter(item => item.trim().length > 0);
                    if (items.length > 0) {
                        return items.map(item => item.trim());
                    }
                }
            }
            
            // Si aucun pattern ne match, chercher les avantages courants dans le texte
            const benefits = [];
            for (const benefit of commonBenefits) {
                const pattern = new RegExp('\\b' + benefit + '\\b', 'i');
                if (pattern.test(text)) {
                    benefits.push(benefit);
                }
            }
            return benefits.length > 0 ? benefits : ['Not specified'];
        };
        
        // Résultat de l'analyse locale
        const result = {
            title: extractInfo(titlePatterns, 'Unknown Position'),
            company: extractInfo(companyPatterns, 'Unknown Company'),
            location: extractInfo(locationPatterns, 'Unknown Location'),
            contractType: extractInfo(contractPatterns),
            experience: extractInfo(experiencePatterns),
            education: extractInfo(educationPatterns),
            salary: extractInfo(salaryPatterns),
            skills: extractSkills(),
            responsibilities: extractResponsibilities(),
            benefits: extractBenefits()
        };
        
        if (this.config.debug) {
            console.log('Résultat de l\'analyse locale:', result);
        }
        
        return result;
    }
    
    /**
     * Extraire le texte d'un fichier (PDF, DOC, DOCX, TXT)
     * @param {File} file - Fichier à traiter
     * @returns {Promise<string>} - Texte extrait du fichier
     * @private
     */
    async _extractTextFromFile(file) {
        if (this.config.debug) {
            console.log('Extraction du texte du fichier:', file.name);
        }
        
        if (file.type === 'text/plain') {
            return new Promise((resolve, reject) => {
                const reader = new FileReader();
                reader.onload = e => resolve(e.target.result);
                reader.onerror = e => reject(e);
                reader.readAsText(file);
            });
        }
        
        if (file.type === 'application/pdf') {
            // Si le module PDFCleaner est disponible, l'utiliser
            if (window.PDFCleaner && this.config.enablePDFCleaning) {
                try {
                    return await window.PDFCleaner.extractTextFromPDF(file);
                } catch (error) {
                    console.error('Erreur lors de l\'extraction avec PDFCleaner:', error);
                    // Fallback sur l'extraction simple
                    return this._basicExtractTextFromPDF(file);
                }
            } else {
                return this._basicExtractTextFromPDF(file);
            }
        }
        
        // Pour les autres types de fichiers, renvoyer une erreur
        throw new Error(`Type de fichier non supporté: ${file.type}`);
    }
    
    /**
     * Extraction basique de texte d'un PDF (à remplacer par une bibliothèque plus robuste)
     * @param {File} file - Fichier PDF
     * @returns {Promise<string>} - Texte extrait du PDF
     * @private
     */
    _basicExtractTextFromPDF(file) {
        // Cette fonction est un placeholder
        // Dans une implémentation réelle, utiliser pdf.js ou une autre bibliothèque
        return new Promise((resolve) => {
            setTimeout(() => {
                resolve(
                    "Contenu basique extrait du PDF. Pour une extraction complète, " +
                    "veuillez utiliser le service d'API ou activer le nettoyage PDF."
                );
            }, 500);
        });
    }
}

// Exposer la classe globalement
window.JobParserAPI = JobParserAPI;

File: backend/job_parser_api.py
#!/usr/bin/env python3
"""
Job Parser API - Backend pour l'analyse des fiches de poste avec GPT

Ce script fournit une API Flask qui permet d'analyser des fiches de poste
en utilisant GPT pour extraire les informations clés.
"""

import os
import json
import tempfile
from typing import Dict, List, Any, Optional
import logging

# Flask et dépendances
from flask import Flask, request, jsonify
from flask_cors import CORS
import requests
from werkzeug.utils import secure_filename

# Traitement des fichiers
import PyPDF2
import docx
import io

# Configuration du logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger("job_parser_api")

# Création de l'application Flask
app = Flask(__name__)
CORS(app)  # Permettre les requêtes CORS

# Configuration
class Config:
    UPLOAD_FOLDER = tempfile.gettempdir()
    MAX_CONTENT_LENGTH = 5 * 1024 * 1024  # 5 MB max
    ALLOWED_EXTENSIONS = {'pdf', 'docx', 'doc', 'txt'}
    # Clé API OpenAI (à configurer via variable d'environnement)
    OPENAI_API_KEY = os.environ.get("OPENAI_API_KEY", "")
    OPENAI_API_URL = "https://api.openai.com/v1/chat/completions"
    OPENAI_MODEL = "gpt-3.5-turbo"
    DEBUG = os.environ.get("DEBUG", "False").lower() in ("true", "1", "t")

app.config.from_object(Config)

# Vérifier si l'extension du fichier est autorisée
def allowed_file(filename: str) -> bool:
    return '.' in filename and filename.rsplit('.', 1)[1].lower() in Config.ALLOWED_EXTENSIONS

# Extraire le texte d'un fichier PDF
def extract_text_from_pdf(file_path: str) -> str:
    try:
        with open(file_path, 'rb') as file:
            reader = PyPDF2.PdfReader(file)
            text = ""
            for page in reader.pages:
                text += page.extract_text() + "\n"
            return text
    except Exception as e:
        logger.error(f"Erreur lors de l'extraction du texte PDF: {e}")
        return ""

# Extraire le texte d'un fichier DOCX
def extract_text_from_docx(file_path: str) -> str:
    try:
        doc = docx.Document(file_path)
        text = ""
        for paragraph in doc.paragraphs:
            text += paragraph.text + "\n"
        return text
    except Exception as e:
        logger.error(f"Erreur lors de l'extraction du texte DOCX: {e}")
        return ""

# Extraire le texte d'un fichier TXT
def extract_text_from_txt(file_path: str) -> str:
    try:
        with open(file_path, 'r', encoding='utf-8') as file:
            return file.read()
    except UnicodeDecodeError:
        # Essayer avec d'autres encodages si utf-8 échoue
        encodings = ['latin-1', 'cp1252', 'iso-8859-1']
        for encoding in encodings:
            try:
                with open(file_path, 'r', encoding=encoding) as file:
                    return file.read()
            except UnicodeDecodeError:
                continue
    except Exception as e:
        logger.error(f"Erreur lors de l'extraction du texte TXT: {e}")
    return ""

# Analyser le texte de la fiche de poste avec GPT
def analyze_job_text_with_gpt(job_text: str) -> Dict[str, Any]:
    """
    Utilise l'API OpenAI pour analyser une fiche de poste
    et en extraire les informations clés.
    """
    if not Config.OPENAI_API_KEY:
        logger.warning("Clé API OpenAI non configurée. Utilisation du parsing local.")
        return analyze_job_locally(job_text)
    
    # Créer le prompt pour GPT
    prompt = f"""
    Extrais les informations clés de cette fiche de poste pour un parser automatique. 
    Pour chaque champ, cherche les informations pertinentes et les retourne au format précis.
    
    Voici la fiche de poste à analyser:
    {job_text}
    
    Retourne les informations au format JSON suivant:
    {{
        "title": "Titre du poste",
        "company": "Nom de l'entreprise",
        "location": "Localisation du poste",
        "contract_type": "Type de contrat (CDI, CDD, etc.)",
        "skills": ["Compétence 1", "Compétence 2", ...],
        "experience": "Niveau d'expérience requis",
        "education": "Formation requise",
        "salary": "Salaire proposé",
        "responsibilities": ["Responsabilité 1", "Responsabilité 2", ...],
        "benefits": ["Avantage 1", "Avantage 2", ...]
    }}
    
    Si certaines informations ne sont pas trouvées, utilise une valeur par défaut comme une chaîne vide ou un tableau vide.
    Réponds uniquement avec le JSON, sans aucun texte supplémentaire.
    """
    
    try:
        # Appel à l'API OpenAI
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {Config.OPENAI_API_KEY}"
        }
        
        payload = {
            "model": Config.OPENAI_MODEL,
            "messages": [
                {"role": "system", "content": "Tu es un assistant spécialisé dans l'analyse de fiches de poste. Tu réponds uniquement au format JSON."},
                {"role": "user", "content": prompt}
            ],
            "temperature": 0.2
        }
        
        response = requests.post(
            Config.OPENAI_API_URL,
            headers=headers,
            json=payload
        )
        
        if response.status_code != 200:
            logger.error(f"Erreur API OpenAI: {response.status_code} - {response.text}")
            return analyze_job_locally(job_text)
        
        # Parser la réponse
        response_data = response.json()
        if "choices" in response_data and len(response_data["choices"]) > 0:
            content = response_data["choices"][0]["message"]["content"]
            
            # Extraire le JSON de la réponse
            try:
                # Tenter d'extraire le JSON de la réponse s'il est entouré de ```
                json_match = content.strip()
                if json_match.startswith("```json"):
                    json_match = json_match[7:]
                if json_match.endswith("```"):
                    json_match = json_match[:-3]
                
                result = json.loads(json_match.strip())
                logger.info("Analyse GPT réussie")
                return result
            except json.JSONDecodeError as e:
                logger.error(f"Erreur lors du parsing JSON de la réponse GPT: {e}")
                logger.debug(f"Contenu reçu: {content}")
                return analyze_job_locally(job_text)
        
        # En cas de problème avec la réponse
        logger.warning("Format de réponse OpenAI inattendu")
        return analyze_job_locally(job_text)
    
    except Exception as e:
        logger.error(f"Erreur lors de l'appel à l'API OpenAI: {e}")
        return analyze_job_locally(job_text)

# Fonction de fallback: analyse locale de la fiche de poste
def analyze_job_locally(job_text: str) -> Dict[str, Any]:
    """
    Analyse la fiche de poste localement avec des expressions régulières.
    Utilisé comme fallback si l'API OpenAI n'est pas disponible.
    """
    import re
    
    # Initialiser l'objet de résultat
    result = {
        "title": "",
        "company": "",
        "location": "",
        "contract_type": "",
        "skills": [],
        "experience": "",
        "education": "",
        "salary": "",
        "responsibilities": [],
        "benefits": []
    }
    
    try:
        # Extraction du titre du poste
        title_patterns = [
            r"(?:^|\n)[\s•]*Poste[\s:]*(.+?)(?:\n|$)",
            r"(?:^|\n)[\s•]*Intitulé du poste[\s:]*(.+?)(?:\n|$)",
            r"(?:^|\n)[\s•]*Offre d'emploi[\s:]*(.+?)(?:\n|$)",
            r"(?:^|\n)[\s•]*([\w\s\-']+(?:développeur|ingénieur|technicien|consultant|manager|responsable|directeur|analyste)[\w\s\-']+)(?:\n|$)"
        ]
        
        for pattern in title_patterns:
            match = re.search(pattern, job_text, re.IGNORECASE)
            if match:
                result["title"] = match.group(1).strip()
                break
        
        # Extraction de l'entreprise
        company_patterns = [
            r"(?:^|\n)[\s•]*Entreprise[\s:]*(.+?)(?:\n|$)",
            r"(?:^|\n)[\s•]*Société[\s:]*(.+?)(?:\n|$)",
            r"(?:^|\n)[\s•]*Employeur[\s:]*(.+?)(?:\n|$)"
        ]
        
        for pattern in company_patterns:
            match = re.search(pattern, job_text, re.IGNORECASE)
            if match:
                result["company"] = match.group(1).strip()
                break
        
        # Extraction de la localisation
        location_patterns = [
            r"(?:^|\n)[\s•]*Lieu[\s:]*(.+?)(?:\n|$)",
            r"(?:^|\n)[\s•]*Localisation[\s:]*(.+?)(?:\n|$)",
            r"(?:^|\n)[\s•]*Localité[\s:]*(.+?)(?:\n|$)",
            r"(?:^|\n)[\s•]*Ville[\s:]*(.+?)(?:\n|$)"
        ]
        
        for pattern in location_patterns:
            match = re.search(pattern, job_text, re.IGNORECASE)
            if match:
                result["location"] = match.group(1).strip()
                break
        
        # Extraction du type de contrat
        contract_patterns = [
            r"(?:^|\n)[\s•]*Type de contrat[\s:]*(.+?)(?:\n|$)",
            r"(?:^|\n)[\s•]*Contrat[\s:]*(.+?)(?:\n|$)",
            r"(?:^|\n)[\s•]*(CDI|CDD|Stage|Alternance|Intérim|Freelance)(?:\n|$)"
        ]
        
        for pattern in contract_patterns:
            match = re.search(pattern, job_text, re.IGNORECASE)
            if match:
                result["contract_type"] = match.group(1).strip()
                break
        
        # Et ainsi de suite pour les autres champs...
        # Le code complet serait trop long, cette version simplifiée extrait juste les principaux champs
        
        logger.info("Analyse locale réussie")
        return result
    
    except Exception as e:
        logger.error(f"Erreur lors de l'analyse locale: {e}")
        return result

# Route pour l'analyse de fichier/texte
@app.route('/api/parse-job', methods=['POST'])
def parse_job():
    """
    Endpoint pour analyser une fiche de poste.
    Accepte soit un fichier, soit un texte brut.
    """
    try:
        # Vérifier si un fichier est fourni
        if 'file' in request.files:
            file = request.files['file']
            
            # Vérifier que le fichier est valide
            if file.filename == '':
                return jsonify({"error": "Nom de fichier vide"}), 400
            
            if not file or not allowed_file(file.filename):
                return jsonify({"error": "Type de fichier non autorisé"}), 400
            
            # Sauvegarder le fichier temporairement
            filename = secure_filename(file.filename)
            file_path = os.path.join(app.config['UPLOAD_FOLDER'], filename)
            file.save(file_path)
            
            # Extraire le texte selon le type de fichier
            if filename.endswith('.pdf'):
                job_text = extract_text_from_pdf(file_path)
            elif filename.endswith(('.docx', '.doc')):
                job_text = extract_text_from_docx(file_path)
            elif filename.endswith('.txt'):
                job_text = extract_text_from_txt(file_path)
            else:
                return jsonify({"error": "Format de fichier non pris en charge"}), 400
            
            # Supprimer le fichier temporaire
            os.remove(file_path)
            
        # Sinon, vérifier si un texte est fourni
        elif 'text' in request.form:
            job_text = request.form['text']
        else:
            return jsonify({"error": "Aucun fichier ou texte fourni"}), 400
        
        # Vérifier que le texte n'est pas vide
        if not job_text or len(job_text.strip()) == 0:
            return jsonify({"error": "Le texte extrait est vide"}), 400
        
        # Analyser le texte avec GPT
        result = analyze_job_text_with_gpt(job_text)
        
        # Retourner le résultat
        return jsonify(result)
    
    except Exception as e:
        logger.exception("Erreur lors du traitement de la requête")
        return jsonify({"error": str(e)}), 500

# Route pour la santé de l'API
@app.route('/api/health', methods=['GET'])
def health_check():
    """Endpoint pour vérifier que l'API est fonctionnelle."""
    return jsonify({"status": "ok", "version": "1.0.0"})

# Lancement de l'application
if __name__ == '__main__':
    # Vérifier si la clé API est configurée
    if not Config.OPENAI_API_KEY:
        logger.warning("Aucune clé API OpenAI n'a été configurée. Certaines fonctionnalités seront limitées.")
    
    # Démarrer le serveur
    port = int(os.environ.get("PORT", 5055))
    app.run(host='0.0.0.0', port=port, debug=Config.DEBUG)
